modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosInstrumentActions.scala

```diff
@@ -78,130 +97,128 @@ class GmosInstrumentActions[F[_]: MonadError[?[_], Throwable]: Logger, A <: Gmos
+      .flatten ++
+      Stream.emit[F, Result[F]](Result.Partial(NSComplete))
 
-  // TODO reduce duplication with respect to InstrumentActions.safeObserve
-  private def safeObserve(
-    i: Int,
+  private def continueObserve(
     env: ObserveEnvironment[F]
   ): Stream[F, Result[F]] =
-    Stream.eval(FileIdProvider.fileId(env)).flatMap { fileId =>
-      observe(i, env)(fileId)
-    }
+    // Steps in between do a continue
+    Stream
+      .eval(
+        inst
+          .calcObserveTime(env.config)
+          .flatMap(inst.continueCommand)
+      )
+      .as(Result.Partial(NSContinue))
+
+  // Calculate the subexposures
+  private def subexposures(
+    cycles: Int
+  ): List[Subexposure] =
+    (for {
+      i     <- 0 until cycles
+      j     <- 0 until Gmos.NsSequence.length
+      stage = Gmos.NsSequence.toList.lift(j).getOrElse(StageA)
+    } yield Subexposure(cycles, i, j, stage)).toList
 
   /**
-   * Calculate the actions for a N&S cycle
-   */
-  private def actionPositions(
-    env:   ObserveEnvironment[F],
-    rows:  Int,
+    * Stream of actions of one sub exposure
+    */
+  def oneSubExposure(
+    fileId:    ImageFileId,
+    sub:       Subexposure,
+    rows:      Int,
     positions: Vector[NSPosition],
-    post:  (Stream[F, Result[F]], ObserveEnvironment[F]) => Stream[F, Result[F]]
-  ): List[ParallelActions[F]] =
-    Gmos.NsSequence.zipWithIndex
-      .map {
-        case (stage, i) =>
-          val rowsToShuffle = if (stage === StageA) 0 else rows
-          val nsPositionO = positions.find(_.stage === stage)
-          List(
-            // Configure rows to shuffle
-            NonEmptyList(
-              inst
-                .configureShuffle(rowsToShuffle)
-                .as(Response.Configured(inst.resource))
-                .toAction(ActionType.Configure(inst.resource)),
-              (env.getTcs, nsPositionO).mapN{ case (tcs, nsPos) =>
-                tcs.nod(
-                  stage,
-                  InstrumentOffset(
-                    tag[OffsetP](nsPos.offset.p.toRadians.radians), tag[OffsetQ](nsPos.offset.q.toRadians.radians)),
-                  nsPos.guide === Guiding.Guide
-                ).toAction(ActionType.Configure(tcs.resource))
-              }.toList
-            ),
-            // Do an obs observe/continue
-            NonEmptyList.one(
-              Action(
-                ActionType.Observe,
-                post( // post lets upstream to mix progress events
-                  safeObserve(i, env),
-                  env
+    env:       ObserveEnvironment[F],
+    post:      (Stream[F, Result[F]], ObserveEnvironment[F]) => Stream[F, Result[F]]
+  ): Stream[F, Result[F]] = {
+    val rowsToShuffle = if (sub.stage === StageA) 0 else rows
+    val nsPositionO   = positions.find(_.stage === sub.stage)
+    // Configure GMOS rows
+    Stream.eval(
+      inst.configureShuffle(rowsToShuffle).as(Result.Partial(NSRowsConfigure))
+    ).mergeHaltBoth(
```

Comment:
Should be `merge`, or you will get only one result. And what is worst, the subexposure will start after the first one finishes.

Reply:
I don't think that will happen, but it could start the next subexposure before the nod completes as `inst.configureShuffle` maybe faster that the nod
I changed it to `merge`